src/ngraph/attribute_visitor.hpp

```diff
@@ -27,137 +28,125 @@ namespace ngraph
-                                ValueAccessor<std::vector<uint32_t>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
+                                ValueAccessor<std::vector<uint32_t>>& adapter);
         virtual void on_adapter(const std::string& name,
-                                ValueAccessor<std::vector<uint64_t>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
-        virtual void on_adapter(const std::string& name, ValueAccessor<std::vector<float>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
+                                ValueAccessor<std::vector<uint64_t>>& adapter);
         virtual void on_adapter(const std::string& name,
-                                ValueAccessor<std::vector<double>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
+                                ValueAccessor<std::vector<float>>& adapter);
         virtual void on_adapter(const std::string& name,
-                                ValueAccessor<std::vector<std::string>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
-        // Use an adapter for non-primitive types
-        template <typename T>
-        // typename std::enable_if<std::is_class<T>::value, void>::type
-        void on_attribute(const std::string& name, T& value)
-        {
-            AttributeAdapter<T> adapter(value);
-            on_adapter(name, adapter);
-        }
-        void on_attribute(const std::string& name, op::AutoBroadcastSpec& value)
-        {
-            AttributeAdapter<op::AutoBroadcastType> adapter(value.m_type);
-            on_adapter(name, adapter);
-        }
-        void on_attribute(const std::string& name, op::BroadcastModeSpec& value)
-        {
-            AttributeAdapter<op::BroadcastType> adapter(value.m_type);
-            on_adapter(name, adapter);
-        }
+                                ValueAccessor<std::vector<double>>& adapter);
+        virtual void on_adapter(const std::string& name,
+                                ValueAccessor<std::vector<std::string>>& adapter);
+        /// \brief Hook for adapters that need visitor access
+        virtual void on_adapter(const std::string& name, VisitorAdapter& adapter);
+
+        /// The generic visitor. There must be a definition of AttributeAdapter<T> that can convert
+        /// to a ValueAccessor<U> for one of the on_adpater methods.
+        template <typename AT>
+        void on_attribute(const std::string& name, AT& value)
+        {
+            AttributeAdapter<AT> adapter(value);
+            on_adapter(get_name_with_context(name), adapter);
+        }
+        /// \returns The nested context of visits
+        const std::vector<std::string>& get_context() const { return m_context; }
```

Comment:
Context, what is it?

Reply:
Suppose an X op has a `before_pad_spec` and an `after_pad_spec` and both are some struct `XPadSpec`. There'd be `visitor.on_attribute("before_pad_spec", m_before_pad_spec")` and that would get to the pad spec adapter which would do `visitor.on_fill("fill", m_fill")` and `visitor.on_size("size", m_size)`. The context would be `["before_pad_spec"]`. When we get to `after_pad_spec` the context would be `["after_pad_spec"]`.  If there were a vector of pad specs, you could have contexts like `["before_pad_spec", "0"]`, `["before_pad_spec", "1"]`, etc. We need to make sure however contexts work matches up with however your XML reading works since I think the XML format is not something that can be changed.